pkg/controller/populators/populator-base.go

```diff
@@ -0,0 +1,208 @@
+	annotations[annPopulationTargetPVC] = pvc.Name
+	annotations[cc.AnnImmediateBinding] = ""
+	if waitForFirstConsumer {
+		annotations[annSelectedNode] = pvc.Annotations[annSelectedNode]
+	}
+	pvcPrimeName := PVCPrimeName(pvc)
+	pvcPrime := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        pvcPrimeName,
+			Namespace:   pvc.Namespace,
+			Labels:      labels,
+			Annotations: annotations,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes:      pvc.Spec.AccessModes,
+			Resources:        pvc.Spec.Resources,
+			StorageClassName: pvc.Spec.StorageClassName,
+			VolumeMode:       pvc.Spec.VolumeMode,
+		},
+	}
+	pvcPrime.OwnerReferences = []metav1.OwnerReference{
+		*metav1.NewControllerRef(pvc, schema.GroupVersionKind{
+			Group:   "",
+			Version: "v1",
+			Kind:    "PersistentVolumeClaim",
+		}),
+	}
+	util.SetRecommendedLabels(pvcPrime, r.installerLabels, "cdi-controller")
+
+	if !isPVCOwnedByDataVolume(pvc) {
+		requestedSize, _ := pvc.Spec.Resources.Requests[corev1.ResourceStorage]
+		// disk or image size, inflate it with overhead
+		requestedSize, err := cc.InflateSizeWithOverhead(r.client, requestedSize.Value(), &pvc.Spec)
+		if err != nil {
+			return nil, err
+		}
+		pvcPrime.Spec.Resources.Requests[corev1.ResourceStorage] = requestedSize
+
+	}
+	if pvcModifier != nil {
+		pvcModifier(pvcPrime)
+	}
+	if err := r.client.Create(context.TODO(), pvcPrime); err != nil {
+		return nil, err
+	}
+
+	r.recorder.Eventf(pvc, corev1.EventTypeNormal, createdPVCPrimeSucceesfully, createdPVCPrimeSucceesfully)
+	return pvcPrime, nil
+}
+
+func (r *ReconcilerBase) rebindPV(pvcPrime *corev1.PersistentVolumeClaim) error {
+	pv, err := r.getPV(pvcPrime.Spec.VolumeName)
+	if pv == nil {
+		return err
+	}
+
+	// Examine the claimref for the PV and see if it's still bound to PVC'
+	claimRef := pv.Spec.ClaimRef
+	if claimRef.Name != pvcPrime.Name || claimRef.Namespace != pvcPrime.Namespace || claimRef.UID != pvcPrime.UID {
+		return nil
+	}
+
+	// Rebind PVC to target PVC
+	targetPVC := &corev1.PersistentVolumeClaim{}
+	targetPVCName, _ := pvcPrime.Annotations[annPopulationTargetPVC]
+	targetPVCKey := types.NamespacedName{Namespace: pvcPrime.Namespace, Name: targetPVCName}
+	if err := r.client.Get(context.TODO(), targetPVCKey, targetPVC); err != nil {
+		return err
+	}
+	// Update PV with strategic values to perform the PV rebind
+	pv.Annotations = make(map[string]string)
+	pv.Spec.ClaimRef = &corev1.ObjectReference{
+		Namespace:       targetPVC.Namespace,
+		Name:            targetPVCName,
+		UID:             targetPVC.UID,
+		ResourceVersion: targetPVC.ResourceVersion,
+	}
+	if err := r.client.Update(context.TODO(), pv); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func addCommonPopulatorsWatches(mgr manager.Manager, c controller.Controller) error {
+	// Setup watches
+	if err := c.Watch(&source.Kind{Type: &corev1.PersistentVolumeClaim{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
```

Comment:
what about when we watch the pod or the PV, in there you only have information for PVC' how are you queuing the target PVC in this type of watches?

Reply:
Should not care about Pods, also don't think that PVs are important either because the PVC will get updated when binding happens

Reply:
so you say I can remove the watch for pods and pvs then? in that case I guess we can reconcile only the target PVC

Reply:
I guess @mhenriks meant using something like [getDataVolumeOp](https://github.com/kubevirt/containerized-data-importer/blob/bdbf5be21c0d73ce201954231b6b00b0b8528ee8/pkg/controller/datavolume/controller-base.go#L231), so it will reconcile only the relevant PVC according to the controller (upload/import/clone...)

Reply:
Ignore it ^^^, you do it much better.